data_sync/mvds.md

```diff
@@ -0,0 +1,104 @@
+# Minimum Viable Data Synchronization
+
+**DRAFT VERSION 1.0**
+
+*Written by Oskar Thorén oskar@status.im & Dean Eigenmann dean@status.im*
+
+## Table of Contents
+
+1. [Abstract](#abstract)
+2. [Definitions](#definitions)
+3. [Wire Protocol](#wire-protocol)
+    1. [Payloads](#payloads)
+4. [Synchronization](#synchronization)
+    1. [State](#state)
+    2. [Flow](#flow)
+    3. [Retransmission](#retransmission)
+5. [Footnotes](#footnotes)
+
+## Abstract
+
+In this specification we describe a minimum viable protocol for data synchronization inspired by the Bramble Synchronization Protocol<sup>1</sup>. This protocol is built on top of the Bramble Synchronization Protocol and was designed in order to ensure reliable messaging between peers communicating across an unreliable peer-to-peer (P2P) network where they may be unreachable or unresponsive. 
+
+We present a functional specification for future implementation<sup>2</sup> as well as reference simulation data which demonstrates its performance.
+
+## Definitions
+
+| Term | Description |
+|------|-------------|
+| **Peer** | The other nodes that a peer is connected to. |
+
+## Wire Protocol
+
+### Payloads
+
+```protobuf
+message Payload {
+  Ack ack = 1;
+  Offer offer = 2;
+  Request request = 3;
+  repeated Message messages = 4;
+}
+
+message Ack {
+  repeated bytes id = 1;
+}
+
+message Message {
+  bytes group_id = 1;
+  int64 timestamp = 2;
+  bytes body = 3;
+}
+
+message Offer {
+  repeated bytes id = 1;
+}
+
+message Request {
+  repeated bytes id = 1;
+}
+```
+
+## Synchronization
+
+### State
+
+State is kept for any message of the types `OFFER`, `REQUEST` and `MESSAGE` we do not keep states for `ACK` messages as we do not retransmit those periodically. The following information is stored for messages:
+
+ - **Type** - Either `OFFER`, `REQUEST` or `MESSAGE`
+ - **Send Count** - The amount of times a message has been sent to a peer.
+ - **Send Epoch** - The next epoch at which a message can be sent to a peer.
+
+### Flow
+
+A maximum of one payload is sent to peers per epoch, this payload contains all `ACK`, `OFFER`, `REQUEST` and `MESSAGE` messages for the specific peer. Payloads are created in reaction to messages recieved by peers or new messages being sent by a node. 
+
+The following rules dictate how payloads are constructed every epoch:
+
+ - A node initially offers a message, this means an `OFFER` is added to the next payload and the state.
+ - When a node recieves an `OFFER`, a `REQUEST` is added to the next payload and the state. 
+ - When a node recieves a `REQUEST` for a previously sent `OFFER`, the `OFFER` is removed from the state and the corresponding `MESSAGE` is added to the next payload and the state.
+ - When a node receives a `MESSAGE`, the `REQUEST` is removed from the state and an `ACK` is added to the next payload.
+ - When a node receives an `ACK`, the `MESSAGE` is removed from the state.
```

Comment:
Wait, why do we do this? What happens if someone else requests message?

Reply:
The state refers to the sync state which is specific to the node that has ACKd. The message is not removed from storage or other states.

Reply:
Ok, so it is still in sync state just not in that specific row for that peer? Can we clarify this maybe?

Also, is this definitely the right thing to do? Agree we want some kind of GC and resource limits, but the less we delete stuff by default the better IMO

Reply:
I mean why do we need to keep the retransmission information for a peer, if that peer has acknowledged a message and we no longer need to retransmit. The sync state simply keeps `Send_Count` and `Send_Epoch`.